File: cake/javascript/reverse_inplace.js
// arrayOfChars is 0 length, return array.
// arrayOfChars is 1 length, return array.
// arrayOfChars is 2 or more, swap first and last, then go to next character.
// If first and last is the same or first > last return arrayOfChars

function reverse(arrayOfChars) {

  // Reverse the input array of characters in place
  if (arrayOfChars.length <= 1) {
    return arrayOfChars;
  }

  let first = 0
  let last  = arrayOfChars.length - 1

  // ES6
  while (first < last) {
    [arrayOfChars[first], arrayOfChars[last]] = [arrayOfChars[last], arrayOfChars[first]]
    first += 1
    last -= 1
  }
  return arrayOfChars
}

// Tests

let desc = 'empty string';
let input = ''.split('');
reverse(input);
let actual = input.join('');
let expected = '';
assertEqual(actual, expected, desc);

desc = 'single character string';
input = 'A'.split('');
reverse(input);
actual = input.join('');
expected = 'A';
assertEqual(actual, expected, desc);

desc = 'longer string';
input = 'ABCDE'.split('');
reverse(input);
actual = input.join('');
expected = 'EDCBA';
assertEqual(actual, expected, desc);

function assertEqual(a, b, desc) {
  if (a === b) {
    console.log(`${desc} ... PASS`);
  } else {
    console.log(`${desc} ... FAIL: ${a} != ${b}`);
  }
}

File: cake/ruby/reverse_inplace.rb
# My first attempt but it required one temp character
# def reverse!(string)
#
#   # Reverse the input string in place
#   left  = 0
#   right = string.length-1
#   char  = string[left]
#
#   while left < right do
#     string[left]  = string[right]
#     string[right] = char
#     left += 1
#     right -= 1
#     char = string[left]
#   end
#
#   string
# end

def reverse!(str)
  left  = 0
  right = str.length - 1

  while left < right

    # swap characters
    str[left], str[right] = str[right], str[left]

    # move towards middle
    left  += 1
    right -= 1
  end
  str
end

def run_tests
  desc = 'empty string'
  string = ''
  reverse!(string)
  expected = ''
  assert_equal(string, expected, desc)

  desc = 'single character string'
  string = 'A'
  reverse!(string)
  expected = 'A'
  assert_equal(string, expected, desc)

  desc = 'longer string'
  string = 'ABCDE'
  reverse!(string)
  expected = 'EDCBA'
  assert_equal(string, expected, desc)
end

def assert_equal(a, b, desc)
  if a == b
    puts "#{desc} ... PASS"
  else
    puts "#{desc} ... FAIL: #{a} != #{b}"
  end
end

run_tests
